### PR TITLE
Add the possibility to customize Album names by introducing template as new option for Album name.

### LINKED ISSUE #150: Enable makeAlbums to use more than the last component of the folder name
When a lot of photos are being shared, the last component of a file path can collide (e.g. 2018/Christmas with 2019/Christmas). 

GooglePhoto does not support hierarchical structures. Concatenating path elements would probably work for many use case (e.g. 2018_Christmas and 2019_Christmas or 2018/Christmas and 2019/Christmas).

### LINKED ISSUE #157: Add the possibility to customise Album names
**Is your feature request related to a problem? Please describe.**
Currently there are two ways to create Album names: `folderPath` and `folderName`. These options doesn't cover too much use cases. As it was raised at #150 .

**Describe the solution you'd like**
I'd like be able to customise the Album name based on the data I've got and using several functions.

**Describe alternatives you've considered**
The alternative is that the user modifies the folders locally before uploading files.

**Additional context**
Something similar to [Mp3Tag](https://help.mp3tag.de/) scripting.

### PR DESCRIPTION
**Pull request by WACKYprog**
**What issue type does this pull request address?** (keep at least one, remove the others)  
/kind feature

resolves #150
resolves #157 

**Does this pull request add new dependencies?**  
Add the possibility to customize Album names by introducing template as new option for Album name.

A small parser has been implemented to support:

```
Tokens:
  %_folderpath% - full path of the folder containing the file.
  %_directory% - name of the folder containing the file.
  %_parent_directory% - Replaced with the name of the parent folder of the file.
  %_day% - day of the month the file was created (in "DD" format).
  %_month% -month the file was created (in "MM" format).
  %_year% - year the file was created (in "YYYY" format).
  %_time% - time the file was created (in "HH:MM:SS" 24-hour format).
  %_time_en% - time the file was created (in "HH:MM:SS AM/PM" 12-hour format).
               
Funcations:
   $lower(string) - converts the string to lowercase.
   $upper(string) - converts the string to uppercase.
   $sentence(string) - converts the string to sentence case.
   $title(string) - converts the string to title case.
   $regex(string, regex, replacement) - replaces the string with the regex replacement.
   $cutLeft(string, length) - cuts the string from the left.
   $cutRight(string, length) - cuts the string from the right.

Example:	"template:%_directory% - %_month%.%_day%.$cutLeft(%_year%,2)"

```

**What to highlight some difference from proposed spec in #150.**

- I spited date to allow easy region/language specific formatting 
- I added time and time_en to allow for PM AM time format as it is common in UK,US, India etc.
- added more string case conversion and made naming bit more clear

**What to highlight some other decision I made**

- parser do supports recursive options so you can do things like: $upper($cutLeft(%_directory%,5))
- parser support string wrapping in '' so you can do case like   $regex(%_directory%, _, ',') 
- datetime is bases on last file modification time. I will likely put another pull request supporting files exif for datetime you think that could be useful. Also with added support for exif more tags can be added.   

**What else do we need to know?**

### CODE REVIEW

File: internal/config/config_test.go [hunk 1]
```diff
@@ -75,6 +76,9 @@ func TestFromFile(t *testing.T) {
 		{"Should fail if AppAPICredentials are invalid", "testdata/invalid-config/EmptyAppAPICredentials.hjson", "", true},
 		{"Should fail if Jobs is empty", "testdata/invalid-config/NoJobs.hjson", "", true},
 		{"Should fail if Album's format is invalid", "testdata/invalid-config/AlbumBadFormat.hjson", "", true},
+		{"Should fail if Album's format is invalid", "testdata/invalid-config/AlbumBadFormat.hjson", "", true},
+		{"Should fail if Album's name auto mathod is invalid", "testdata/invalid-config/AlbumBadAutoMethod.hjson", "", true},
```
Comment (pacoorozco) [inline internal/config/config_test.go:80]: Typo: `mathod` instead of `method`.

File: internal/config/schema.go [hunk 2]
```diff
@@ -32,11 +32,32 @@ type FolderUploadJob struct {
...
 	//   "auto:" : Followed either "folderPath" or "folderName" will use an autogenerated album name based on the
 	//             object's folder path or object's folder name.
+	//   "template": Followed by a template string that can contain the following predefine tokens and functions:
+	//              Tokens:
+	//                 %_folderpath% - full path of the folder containing the file.
+	//                 %_directory% - name of the folder containing the file.
+	//                 %_parent_directory% - Replaced with the name of the parent folder of the file.
+	//                 %_day% - day of the month the file was created (in "DD" format).
+	//                 %_month% -month the file was created (in "MM" format).
+	//				   %_year% - year the file was created (in "YYYY" format).
```
Comment (pacoorozco) [inline internal/config/schema.go:47]: Do you mind to align these two lines for readability?

File: internal/config/schema.go [hunk 3]
```diff
@@ -32,11 +32,32 @@ type FolderUploadJob struct {
...
+	//				   %_time% - time the file was created (in "HH:MM:SS" 24-hour format).
+	//                 %_time_en% - time the file was created (in "HH:MM:SS AM/PM" 12-hour format).
+	//              Funcations:
```
Comment (pacoorozco) [inline internal/config/schema.go:50]: Typos: `funcations` instead of `functions`

File: internal/config/schema.go [hunk 4]
```diff
@@ -32,11 +32,32 @@ type FolderUploadJob struct {
...
+	//				   $lower(string) - converts the string to lowercase.
+	//				   $upper(string) - converts the string to uppercase.
+	//				   $sentence(string) - converts the string to sentence case.
+	//				   $title(string) - converts the string to title case.
+	//			       $regex(string, regex, replacement) - replaces the string with the regex replacement.
```
Comment (pacoorozco) [inline internal/config/schema.go:55]: There is some miss-alignment in the functions list

File: internal/upload/album_test.go [hunk 5]
```diff
@@ -98,3 +118,121 @@ func TestAlbumNameUsingFolderName(t *testing.T) {
...
+		{in: "cutLeft(", out: ""},
+		{in: "$cutLeft(anything", out: "cutLeft"},
+	}
+
+	for _, tt := range testData {
+		assert.Equal(t, tt.out, getTemplateFunctionName(tt.in))
+	}
+}
+
+func TestParseAlbumNameTampleWithTokens(t *testing.T) {
```
Comment (pacoorozco) [inline internal/upload/album_test.go:139]: Maybe a typo: `Tample` instead of `Template`?

File: internal/upload/album_test.go [hunk 6]
```diff
@@ -98,3 +118,121 @@ func TestAlbumNameUsingFolderName(t *testing.T) {
...
+		val, err := parseAlbumNameTemplate(tt.in, filePath, timeObj)
+		if err != nil {
+			t.Errorf("Unexpected error: %v", err)
+		}
+
+		assert.Equal(t, tt.out, val)
+	}
+}
+
+func TestParseAlbumNameTampleWithFunctions(t *testing.T) {
```
Comment (pacoorozco) [inline internal/upload/album_test.go:170]: Same as above, I'm wondering if this `Tample` is a short name... in that case, feel free to leave is as it is

File: internal/upload/album_test.go [hunk 7]
```diff
@@ -98,3 +118,121 @@ func TestAlbumNameUsingFolderName(t *testing.T) {
...
+		if err != nil {
+			t.Errorf("Unexpected error: %v", err)
+		}
+
+		assert.Equal(t, tt.out, val)
+	}
+}
+
+func TestParseAlbumNameTampleWithInvalidParameter(t *testing.T) {
+	// testing  for Y2k bug ;)
```
Comment (pacoorozco) [inline internal/upload/album_test.go:206]: 😆

**Review by pacoorozco — COMMENTED** [review]
Thank you very much for contributing to this humble project with a feature that had been in the backlog for so long. I think it is something that some people were waiting for. 

I have reviewed the code and it looks very good. Good test coverage and very readable code. Thanks for the effort in implementing the feature and trying to keep the code consistent.

I have made some comments, mostly typos. Once you review them ping me and I'll do a merge as well as a new release of the CLI. 

Once again, thanks!

File: internal/upload/album_test.go (hunk 5 above)
Comment (WACKYprog) [reply internal/upload/album_test.go:139]: yeah will correct

File: internal/config/schema.go (hunk 2 above)
Comment (WACKYprog) [reply internal/config/schema.go:47]: No looks like mixed spaces and tabs and Git hub show tabs as 4. while my setting is showing tabs at 2. Changing to all spaces

**Comment by WACKYprog** [comment]
> Thank you very much for contributing to this humble project with a feature that had been in the backlog for so long. I think it is something that some people were waiting for.
> 
> I have reviewed the code and it looks very good. Good test coverage and very readable code. Thanks for the effort in implementing the feature and trying to keep the code consistent.
> 
> I have made some comments, mostly typos. Once you review them ping me and I'll do a merge as well as a new release of the CLI.
> 
> Once again, thanks!

Happy to help.

I use the project and really have fun working on it. Hopefully will have time to work on it more!

**Review by pacoorozco — APPROVED** [review]
Well done and thanks for addressing all the comment 🙇🏼

**Pull request merged**